File: qsort.js
function RandInt(max){
	return Math.floor(Math.random() * max);
}

function Add(a, b){
	for (let i=0; i<b.length; i++){
		a.push(b[i]);
	}
	return a;
}

function Qsort(a){
	if (a.length <= 1) return a;
	//const x = []; const more = []; const less = [];
	let x = new Array(); let more = new Array(); let less = new Array();
	let piv = a[a.length-1]
	for (let i=0; i<a.length-1; i++){
		if (a[i]>piv) more.push(a[i]);
		else          less.push(a[i]);
	}
	Add(x, (Qsort(less)));
	x.push(piv);
	Add(x, (Qsort(more)));
	return x;
}


//let tab = [3, 5, 4, 2, 1];
let tab = [];

for (let i = 0; i < 100000; i ++){
	tab.push(RandInt(1000000));
}

console.log(tab);
console.log(Qsort(tab));

File: qsort.rs
fn Add(x: &[i32], y: &[i32]) -> Vec<i32>{
	let mut a = x.to_vec();
	for i in 0..y.len(){
		a.push(y[i]);
	}
	return a;
}

fn Qsort(x: &[i32]) -> Vec<i32>{
	let mut y = x;
	if y.len() <= 1 {
		return y.to_vec();
	}
	let mut a = Vec::new();
	let mut m = Vec::new();
	let mut w = Vec::new();
	let piv = y[y.len() - 1];
	for i in 0..y.len(){
		let c = y[i];
		if c > piv {
			w.push(y[i]);
		}
		else if c < piv {
			m.push(y[i]);
		}
	}
	a = Qsort(&m).to_vec();
	a.push(piv);
	a = Add(&a, &Qsort(&w).to_vec());
	return a;
}

fn main(){
	let mut v = Vec::new();
	v.push(3);
	v.push(5);
	v.push(2);
	v.push(1);
	v.push(4);
	let mut w = Qsort(&v).to_vec();
	for i in 0..w.len() {
		println!("{}", w[i]);
	}
	
}
